packages/shared/src/components/squads/SquadJoinButton.tsx

```diff
@@ -0,0 +1,112 @@
+import React, { ReactElement, useEffect } from 'react';
+import classNames from 'classnames';
+import { useMutation, useQueryClient } from 'react-query';
+import { Squad } from '../../graphql/sources';
+import { Button } from '../buttons/Button';
+import { useAuthContext } from '../../contexts/AuthContext';
+import { useToastNotification } from '../../hooks/useToastNotification';
+import { useLeaveSquad, useJoinSquad } from '../../hooks';
+import { labels } from '../../lib';
+import { useAnalyticsContext } from '../../contexts/AnalyticsContext';
+import { AnalyticsEvent, Origin, TargetType } from '../../lib/analytics';
+
+type SquadJoinProps = {
+  className?: string;
+  squad: Squad;
+  joinText?: string;
+  leaveText?: string;
+  origin: Origin;
+};
+
+export const SquadJoinButton = ({
+  className,
+  squad,
+  joinText = 'Join squad',
+  leaveText = 'Leave squad',
+  origin,
+}: SquadJoinProps): ReactElement => {
+  const queryClient = useQueryClient();
+  const { trackEvent } = useAnalyticsContext();
+  const { displayToast } = useToastNotification();
+  const { user, showLogin } = useAuthContext();
+  const isCurrentMember = !!squad.currentMember;
+
+  useEffect(() => {
+    trackEvent({
+      event_name: AnalyticsEvent.Impression,
+      target_type: TargetType.SquadJoinButton,
+      extra: JSON.stringify({
+        squad: squad.id,
+        origin,
+        squad_type: squad.public ? 'public' : 'private',
+      }),
+    });
+    // @NOTE see https://dailydotdev.atlassian.net/l/cp/dK9h1zoM
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [origin, squad.id, squad.public]);
```

Comment:
Not a big fan of passing depedency for the sake of impression calls.
It's important they don't trigger more then once.

I'd either suggest simply removing dependencies here or adding a state to ensure it only fires once per mount.

I know very unlikely it would change, but keep in mind you invalidate queries, so we could accidentally trigger again.

Reply:
But we need those dependencies, if squad id changes or button get's rendered with different squad ID or place (with different origin) we want to trigger impression again right? @rebelchris

Reply:
Which would be a new mount, afaik there is only a danger of retriggering impressions.
CC: @idoshamun Maybe your thoughts?

Reply:
If there's a chance that it will get rendered then it's better to have the deps. Otherwise let's stick to no deps. I think when we discussed with tkdodo he also mentioned tracking as one exception use case for not using the proper deps

Reply:
@capJavert for me this one is still open 🫣

Reply:
I'll just concede for now, let's discuss when we get on implementing special tracking hooks. 🤝